File: apps/api/python/translator_api/models/separation.py
"""Audio separation tracks persistence.

One project can have multiple SeparationTracks per kind (vocals, music, sfx).
Each track row references a storage key in S3 (or local storage).
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from translator_api.models import Base


# Canonical track kinds. Some providers emit more granular (e.g. `vocals`,
# `drums`, `bass`, `other`). We collapse to the four canonical kinds unless
# `extra_kinds` is set on the project.
DEFAULT_KINDS = ("vocals", "music", "sfx", "instrumental")


class SeparationTrack(Base):
    __tablename__ = "separation_tracks"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    project_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False
    )
    kind: Mapped[str] = mapped_column(String(32), nullable=False)
    storage_key: Mapped[str] = mapped_column(String(512), nullable=False)
    provider_id: Mapped[str] = mapped_column(String(64), nullable=False)
    duration_ms: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    sample_rate: Mapped[int] = mapped_column(Integer, nullable=False, default=44100)
    confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

File: apps/web/lib/separation.ts
/** Audio separation API client. */
import { api } from "./api";

export interface SeparationTrack {
  id: string;
  kind: string;
  storage_key: string;
  download_url: string;
  provider_id: string;
  duration_ms: number;
  sample_rate: number;
  confidence: number | null;
  created_at: string;
}

export interface SeparationRunInput {
  provider_id?: string;
  method?: string;
  segment_size?: number;
}

export interface SeparationRunOutput {
  project_id: string;
  provider_id: string;
  method: string;
  tracks: SeparationTrack[];
}

export async function runSeparation(
  projectId: string,
  body: SeparationRunInput,
): Promise<SeparationRunOutput> {
  return api.post<SeparationRunOutput>(`/projects/${projectId}/separation/run`, body);
}

export async function listSeparationTracks(
  projectId: string,
  kind?: string,
): Promise<SeparationTrack[]> {
  const q = kind ? `?kind=${kind}` : "";
  return api.get<SeparationTrack[]>(`/projects/${projectId}/separation/tracks${q}`);
}

export async function deleteSeparationTrack(
  projectId: string,
  trackId: string,
): Promise<void> {
  await api.delete(`/projects/${projectId}/separation/tracks/${trackId}`);
}
